src/realm/util/encrypted_file_mapping.cpp

```diff
@@ -150,6 +150,25 @@ size_t check_read(FileDesc fd, off_t pos, void* dst, size_t len)
     return ret;
 }
 
+inline void memcpy_tsan_safe(char* dst, char* src, size_t num_bytes)
+{
+#if REALM_SANITIZE_THREAD
+            // Copying an entire page may overwrite arrays which is being read concurrently by
+            // another thread. This is benign because it overwrites data from old version with
+            // the same value as is already there, so the reader sees the correct value. This
+            // is all by design. However, when doing a memcpy of the entire page, TSAN sees
```

Comment:
So, do you essentially think that this is just false positive? Which part is "all by design"? The obvious question from my side is why it tries to overwrite with exactly the same data :D

Reply:
@kiburtse only some of the data on the page is expected to change when advancing versions. The parts of the page that are populated with array data for readers of older versions will not actually change, but the parts of the page that were "free space" may now be allocated and filled with new data by an external writer. TSAN says that the memcpy triggers a "write" on the old version data, even if the values didn't change. I was attempting to fix that by only copying the changed bytes, but this strategy doesn't appear to work because we still have to read the entire page to do this, which triggers another separate TSAN race (also possibly a false positive). I'm not sure this strategy is worth pursuing anymore, but I am hesitant to just add a tsan suppression because it may mask real problems.